encoding/tsz/iterator.go

```diff
@@ -280,16 +281,171 @@ func (it *iterator) Reset(reader io.Reader) {
 	it.done = false
 	it.err = nil
 	it.ant = nil
+	it.tu = xtime.None
 	it.closed = false
 }
 
-func (it *iterator) Close() {
+func (it *singleReaderIterator) Close() {
 	if it.closed {
 		return
 	}
 	it.closed = true
-	pool := it.opts.GetIteratorPool()
+	pool := it.opts.GetSingleReaderIteratorPool()
 	if pool != nil {
 		pool.Put(it)
 	}
 }
+
+// An IteratorHeap is a min-heap of iterators. The top of the heap is the iterator
+// whose current value is the earliest datapoint among all iterators in the heap.
+type iteratorHeap []m3db.Iterator
+
+func (h iteratorHeap) Len() int {
+	return len(h)
+}
+
+func (h iteratorHeap) Less(i, j int) bool {
+	di, _, _ := h[i].Current()
+	dj, _, _ := h[j].Current()
+	return di.Timestamp.Before(dj.Timestamp)
+}
+
+func (h iteratorHeap) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+}
+
+func (h *iteratorHeap) Push(x interface{}) {
+	*h = append(*h, x.(m3db.Iterator))
+}
+
+func (h *iteratorHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	if n == 0 {
+		return nil
+	}
+
+	x := old[n-1]
+	*h = old[:n-1]
+	return x
+}
+
+type newSingleReaderIteratorFn func(reader io.Reader, opts Options) m3db.SingleReaderIterator
+
+func getNewSingleReaderIteratorFn(opts Options) newSingleReaderIteratorFn {
```

Comment:
Shouldn't need this method, a pool should always alloc if it is empty and give you back something, so always use a pool with size 0 to avoid this type of checking.